modules/space/horizonsfile.cpp

```diff
@@ -0,0 +1,877 @@
+}
+
+std::pair<std::string, std::string> HorizonsFile::parseValidTimeRange(
+                                                           const std::string& startPhrase,
+                                                             const std::string& endPhrase,
+                                                        const std::string& altStartPhrase,
+                                                                      bool hasTime) const
+{
+    std::ifstream fileStream(_file);
+
+    if (!fileStream.good()) {
+        fileStream.close();
+        return { "", "" };
+    }
+
+    // Ignore everything until head of time range list
+    std::string line;
+    std::getline(fileStream, line);
+    while (fileStream.good()) {
+        // Add the line with the start phrase first, to give context
+        if (line.find(startPhrase) != std::string::npos) {
+            break;
+        }
+        else if (!altStartPhrase.empty() && line.find(altStartPhrase) != std::string::npos) {
+            break;
+        }
+
+        std::getline(fileStream, line);
+    }
+
+    if (!fileStream.good()) {
+        fileStream.close();
+        return { "", "" };
+    }
+
+    // There will be one empty line before the list of time ranges, skip
+    std::getline(fileStream, line);
+
+    // From the first line get the start time
+    std::string startTime, endTime;
+    std::getline(fileStream, line);
+    if (fileStream.good()) {
+        std::stringstream str(line);
+
+        // Read and save each word
+        std::vector<std::string> words;
+        std::string word;
+        while (str >> word) {
+            words.push_back(word);
+        }
+
+        if (words.empty()) {
+            return { "", "" };
+        }
+
+        // Parse time stamps backwards
+        // Format: Trajectory file Name, Start, End (yyyy-mon-dd hh:mm)
+        if (hasTime && words.size() > 4) {
+            startTime = words[words.size() - 4] + " T " + words[words.size() - 3];
+            endTime = words[words.size() - 2] + " T " + words[words.size() - 1];
+        }
+        else if (words.size() > 2){
+            // Sometimes the format can be yyyy-mon-dd without time
+            startTime = words[words.size() - 2];
+            endTime = words[words.size() - 1];
+        }
+        else {
+            return { "", "" };
+        }
+    }
+    if (startTime.empty() || endTime.empty()) {
+        fileStream.close();
+        return { "", "" };
+    }
+
+    // Get the end time from the last trajectery
+    while (fileStream.good()) {
+        if (line.find(endPhrase) != std::string::npos || line.empty() || line == " ") {
+            fileStream.close();
+            return { startTime, endTime };
+        }
+
+        // Read and save each word.
+        std::stringstream str(line);
+        std::vector<std::string> words;
+        std::string word;
+        while (str >> word) {
+            words.push_back(word);
+        }
+
+        if (words.empty()) {
```

Comment:
This check does not seem needed to me. will be caught by the `if {} else if {} else` below?

Reply:
Same on line 814 above? Also ,the parsing up there seems very similar to how it's done here. Can it be generalized?

Reply:
It cant really be generalized since there are som small differences between the two cases. But I added a longer explanation before the function to make it clearer what the difference is.